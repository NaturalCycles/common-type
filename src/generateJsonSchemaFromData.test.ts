import { _range } from '@naturalcycles/js-lib'
import { AjvSchema } from '@naturalcycles/nodejs-lib'
import { generateJsonSchemaFromData } from './generateJsonSchemaFromData'

test('generateJsonSchemaFromData1', () => {
  const data1 = _range(5).map(i => {
    return { a: 'a', n: i, b: i % 2 === 0, c: null, u: undefined }
  })

  const s = generateJsonSchemaFromData(data1)
  // console.log(s)
  expect(s).toMatchInlineSnapshot(`
    Object {
      "additionalProperties": true,
      "properties": Object {
        "a": Object {
          "type": "string",
        },
        "b": Object {
          "type": "boolean",
        },
        "c": Object {
          "type": "null",
        },
        "n": Object {
          "type": "number",
        },
      },
      "required": Array [],
      "type": "object",
    }
  `)

  const schema = AjvSchema.create(s)
  data1.forEach(r => schema.validate(r))
})

test('generateJsonSchemaFromData2', () => {
  const data2 = _range(20).map(i => {
    const r: any = { a: 'a', n: i, b: i % 2 === 0, u: undefined }

    if (i % 3 === 0) {
      r.c = null

      r.sub = {
        aa: 'aa',
        sub2: {
          n: null,
          b: false,
        },
      }
    }

    if (i % 4 === 0) {
      r.sub = false
    }

    if (i % 5 === 0) {
      r.arr = [1, 2, 3]
      r.arr2 = [1, 2, 3, 'str', true, { a: 's' }]
    }

    return r
  })

  const s = generateJsonSchemaFromData(data2)

  // console.log(JSON.stringify(s, null, 2))
  // console.log(inspectAny(s, {
  //   depth: 100,
  //   maxLen: 1000000,
  // }))
  expect(s).toMatchSnapshot()

  const schema = AjvSchema.create(s)
  data2.forEach(r => schema.validate(r))
})
