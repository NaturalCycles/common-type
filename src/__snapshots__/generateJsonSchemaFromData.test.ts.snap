// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateJsonSchemaFromData2 1`] = `
Object {
  "additionalProperties": true,
  "properties": Object {
    "a": Object {
      "type": "string",
    },
    "arr": Object {
      "items": Object {
        "type": "number",
      },
      "type": "array",
    },
    "arr2": Object {
      "items": Object {
        "oneOf": Array [
          Object {
            "type": "number",
          },
          Object {
            "type": "string",
          },
          Object {
            "type": "boolean",
          },
          Object {
            "additionalProperties": true,
            "properties": Object {
              "a": Object {
                "type": "string",
              },
            },
            "required": Array [],
            "type": "object",
          },
        ],
      },
      "type": "array",
    },
    "b": Object {
      "type": "boolean",
    },
    "c": Object {
      "type": "null",
    },
    "n": Object {
      "type": "number",
    },
    "sub": Object {
      "oneOf": Array [
        Object {
          "type": "boolean",
        },
        Object {
          "additionalProperties": true,
          "properties": Object {
            "aa": Object {
              "type": "string",
            },
            "sub2": Object {
              "additionalProperties": true,
              "properties": Object {
                "b": Object {
                  "type": "boolean",
                },
                "n": Object {
                  "type": "null",
                },
              },
              "required": Array [],
              "type": "object",
            },
          },
          "required": Array [],
          "type": "object",
        },
      ],
    },
  },
  "required": Array [],
  "type": "object",
}
`;
